lib/trino-hive-formats/src/main/java/io/trino/hive/formats/rcfile/RcFileReader.java

```diff
@@ -144,100 +143,119 @@ private RcFileReader(
+            columns = new Column[columnCount];
+            for (Entry<Integer, Type> entry : readColumns.entrySet()) {
+                if (entry.getKey() < columnCount) {
+                    ColumnEncoding columnEncoding = encoding.getEncoding(entry.getValue());
+                    columns[entry.getKey()] = new Column(columnEncoding, decompressor);
+                }
+            }
 
-        // initialize columns
-        verify(columnCount <= MAX_COLUMN_COUNT, "Too many columns (%s) in RCFile %s", columnCountString, inputFile.location());
-        columns = new Column[columnCount];
-        for (Entry<Integer, Type> entry : readColumns.entrySet()) {
-            if (entry.getKey() < columnCount) {
-                ColumnEncoding columnEncoding = encoding.getEncoding(entry.getValue());
-                columns[entry.getKey()] = new Column(columnEncoding, decompressor);
+            // read sync bytes
+            syncFirst = input.readLong();
+            validateWrite(validation -> validation.getSyncFirst() == syncFirst, "Unexpected sync sequence");
+            syncSecond = input.readLong();
+            validateWrite(validation -> validation.getSyncSecond() == syncSecond, "Unexpected sync sequence");
+
+            // seek to first sync point within the specified region, unless the region starts at the beginning
+            // of the file.  In that case, the reader owns all row groups up to the first sync point.
+            if (offset != 0) {
+                // if the specified file region does not contain the start of a sync sequence, this call will close the reader
+                long startOfSyncSequence = ReadWriteUtils.findFirstSyncPosition(inputFile, offset, length, syncFirst, syncSecond);
+                if (startOfSyncSequence < 0) {
+                    closeQuietly();
+                    return;
+                }
+                input.seek(startOfSyncSequence);
             }
         }
-
-        // read sync bytes
-        syncFirst = input.readLong();
-        validateWrite(validation -> validation.getSyncFirst() == syncFirst, "Unexpected sync sequence");
-        syncSecond = input.readLong();
-        validateWrite(validation -> validation.getSyncSecond() == syncSecond, "Unexpected sync sequence");
-
-        // seek to first sync point within the specified region, unless the region starts at the beginning
-        // of the file.  In that case, the reader owns all row groups up to the first sync point.
-        if (offset != 0) {
-            // if the specified file region does not contain the start of a sync sequence, this call will close the reader
-            long startOfSyncSequence = ReadWriteUtils.findFirstSyncPosition(inputFile, offset, length, syncFirst, syncSecond);
-            if (startOfSyncSequence < 0) {
-                closeQuietly();
-                return;
+        catch (Throwable e) {
```

Comment:
Also please use this cleaner idom:
```
        catch (Throwable throwable) {
            try (input) {
                throw throwable;
            }
        }
```

Reply:
See [SequenceFileReader](https://github.com/trinodb/trino/blob/a74f29d1fa219b3b4cd3d568b257b035ef496741/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/sequence/SequenceFileReader.java#L167) for a full example.  SequenceFileReader is very new and uses more modern techniques

Reply:
This looks better. I can just copy the logic in SequenceFileReader over.